### PR TITLE
Reworked Miner Logging

### PR DESCRIPTION
**Pull request by fosskers**
This PR reworks the logging in `chainweb-miner`, giving it timestamps and proper configurable log levels. It also cleans up the Miner code in general and fixes a bug.

2019-09-15-170014_958x538_scrot

### CODE REVIEW

File: chainweb.cabal [hunk 1]
```diff
@@ -539,32 +538,22 @@ executable chainweb-miner
...
-        , containers >= 0.5
         , errors >= 2.3
+        , generic-lens >= 1.2
         , http-client >= 0.5
         , http-client-tls >= 0.3
         , mwc-random >= 0.13
         , optparse-applicative >= 0.14
-        , pretty-simple >= 2.1
         , retry >= 0.7
+        , rio >= 0.1.9
```
Comment (fosskers) [inline chainweb.cabal]: RIO offered the simplest logging infrastructure that I knew how to get running quickly. Its logging has nice looking, reasonable defaults. It cleans up a lot of code below, too.

File: miner/Miner.hs [hunk 2]
```diff
@@ -59,91 +56,67 @@
...
-    , mgr :: Manager
-    , args :: ClientArgs }
+    { gen :: !MWC.GenIO
+    , mgr :: !Manager
+    , log :: !LogFunc
+    , args :: !ClientArgs }
+    deriving stock (Generic)
+
+instance HasLogFunc Env where
+    logFuncL = field @"log"
```
Comment (fosskers) [inline miner/Miner.hs]: The only boilerplate that RIO demands for its logging.

File: miner/Miner.hs [hunk 3]
```diff
@@ -208,107 +190,114 @@ main = do
...
-    CPU e -> cpu (version $ args env) e (gen env)
-    GPU _ -> gpu
+run :: RIO Env ()
+run = do
+    logInfo "Starting Miner."
+    env <- ask
+    case cmd $ args env of
+        GPU -> logError "GPU mining is not yet available."
+        CPU _ -> getWork >>= traverse_ (mining (scheme env))
+    exitFailure
```
Comment (fosskers) [inline miner/Miner.hs]: The miner is designed to never exit. If `mining` ever returns, we should consider it a failure, hence `exitFailure`.

File: miner/Miner.hs [hunk 4]
```diff
@@ -208,107 +190,114 @@ main = do
     opts = info (pClientArgs <**> helper)
         (fullDesc <> progDesc "The Official Chainweb Mining Client")
 
-scheme :: Env -> (TargetBytes -> HeaderBytes -> IO HeaderBytes)
-scheme env = case cmd $ args env of
-    CPU e -> cpu (version $ args env) e (gen env)
-    GPU _ -> gpu
+run :: RIO Env ()
```
Comment (fosskers) [inline miner/Miner.hs]: Effectively `ReaderT` with built-in support for logging. Mostly a drop-in replacement.

File: miner/Miner.hs [hunk 5]
```diff
@@ -208,107 +190,114 @@ main = do
...
     -- won the race instead, then the `go` call is automatically cancelled.
     --
-    miningSuccess :: HeaderBytes -> IO ()
+    miningSuccess :: HeaderBytes -> RIO Env ()
     miningSuccess h = do
-      bh <- bytesToBlockHeader h
-      putStrLn $ "Success!!!  " <> (show $ _blockNonce bh)
-      putStrLn $ "difficulty = " <> (T.unpack $ showTargetHex $ _blockTarget bh)
-      putStrLn $ "blockHash  = " <> (T.unpack $ hashToHex $ _blockHash bh)
-      putStrLn $ "powHash    = " <> (T.unpack $ powHashToHex $ _blockPow bh)
```
Comment (fosskers) [inline miner/Miner.hs]: @mightybyte we can reintroduce these if you like, possibly at the `Debug` level. The day-1 user probably just wants to know if they mined successfully or not, which the `logInfo` call below does.

File: miner/Miner.hs [hunk 6]
```diff
@@ -208,107 +190,114 @@ main = do
...
-bytesToBlockHeader (HeaderBytes hbytes) = runGet decodeBlockHeaderWithoutHash hbytes
-
-cpu :: ChainwebVersion -> CPUEnv -> MWC.GenIO -> TargetBytes -> HeaderBytes -> IO HeaderBytes
-cpu v e g tbytes hbytes = fmap head . withScheduler comp $ \sch -> do
-    bh <- bytesToBlockHeader hbytes
-    let BlockHeight height = _blockHeight bh
-    printf "Mining block %d on chain %s with difficulty %s\n"
-      height
-      (T.unpack $ chainIdToText $ _blockChainId bh)
-      (T.unpack $ showTargetHex (_blockTarget bh))
```
Comment (fosskers) [inline miner/Miner.hs]: Doing all these here would slow down mining.

File: chainweb.cabal [hunk 7]
```diff
@@ -539,32 +538,22 @@ executable chainweb-miner
...
-        , containers >= 0.5
         , errors >= 2.3
+        , generic-lens >= 1.2
         , http-client >= 0.5
         , http-client-tls >= 0.3
         , mwc-random >= 0.13
         , optparse-applicative >= 0.14
-        , pretty-simple >= 2.1
         , retry >= 0.7
+        , rio >= 0.1.5
```
Comment (fosskers) [inline chainweb.cabal]: RIO offered the simplest logging infrastructure that I knew how to get running quickly. Its logging has nice looking, reasonable defaults. It cleans up a lot of code below, too.

File: miner/Miner.hs [hunk 8]
```diff
@@ -208,107 +191,114 @@ main = do
...
-    CPU e -> cpu (version $ args env) e (gen env)
-    GPU _ -> gpu
+run :: RIO Env ()
+run = do
+    logInfo "Starting Miner."
+    env <- ask
+    case cmd $ args env of
+        GPU -> logError "GPU mining is not yet available."
+        CPU _ -> getWork >>= traverse_ (mining (scheme env))
+    liftIO exitFailure
```
Comment (fosskers) [inline miner/Miner.hs]: The miner is designed to never exit. If mining ever returns, we should consider it a failure, hence `exitFailure`.

File: miner/Miner.hs [hunk 9]
```diff
@@ -188,15 +174,12 @@ pChainId = optional $ textOption
...
-        <$> MWC.createSystemRandom
-        <*> newManager (mkManagerSettings ss Nothing)
-        <*> execParser opts
-    getWork env >>= \case
-        Nothing -> putStrLn "Failed to connect to the given Chainweb Node." >> exitFailure
-        Just wb -> case cmd as of
-            GPU _ -> putStrLn "GPU mining is not yet available."
-            CPU _ -> pPrintNoColor as >> mining (scheme env) env wb >> exitFailure
+    cargs <- execParser opts
+    lopts <- setLogMinLevel (ll cargs) . setLogUseLoc False <$> logOptionsHandle stderr True
```
Comment (fosskers) [inline miner/Miner.hs]: "Be verbose, but set the minimum level to whatever came in from the CLI, and don't bother echoing source line locations."

File: chainweb.cabal (hunk 7 above)
Comment (mightybyte) [reply chainweb.cabal]: Yet another logging system?  Not a fan of that.  Also, rio seems like overkill here.  Let's stick with the logging system we've already got.  Either that or print statements are perfectly fine.

File: miner/Miner.hs [hunk 10]
```diff
@@ -208,107 +191,114 @@ main = do
...
-scheme :: Env -> (TargetBytes -> HeaderBytes -> IO HeaderBytes)
-scheme env = case cmd $ args env of
-    CPU e -> cpu (version $ args env) e (gen env)
-    GPU _ -> gpu
+run :: RIO Env ()
+run = do
+    logInfo "Starting Miner."
+    env <- ask
+    case cmd $ args env of
+        GPU -> logError "GPU mining is not yet available."
```
Comment (mightybyte) [inline miner/Miner.hs]: If it's not working, we should take this out of the command-line help.  We shouldn't ship non-working features.

File: miner/Miner.hs (hunk 6 above)
Comment (mightybyte) [reply miner/Miner.hs]: This is a few instructions once every time a new block gets mined.  This is so far down in the weeds it's not worth worrying about.  Also, the instructions for doing these conversions will be dwarfed by the IO itself.

File: miner/Miner.hs (hunk 5 above)
Comment (mightybyte) [reply miner/Miner.hs]: I think day-1 users will probably want to know how to find the block that they mined.  This means at least showing them chain ID and block height.  Also probably block hash.  But then people expect to see zeroes in the block hash, so `powHash` makes sense as well.  Difficulty we could probably leave out.  That was more for my benefit as a sanity check.

File: miner/Miner.hs [hunk 11]
```diff
@@ -208,107 +191,114 @@ main = do
...
-        -- there won't be any overlap?
-        n <- Nonce <$> MWC.uniform g
-        new <- usePowHash v (\p -> mine p n tbytes) hbytes
-        terminateWith sch new
+      e <- ask
+      let !v = version $ args e
+          !m = mgr e
+          !u = coordinator $ args e
+      cid <- liftIO chain
+      logInfo $ cid <> ": Mining Success!"
```
Comment (mightybyte) [inline miner/Miner.hs]: I think we need to tell people what that first number is.  They won't magically know.  Also, we should at least tell them the block height of the block they just mined.  That gives the minimum amount of information they need to go look the block up using the chainweb API or in the future a block explorer.

File: miner/Miner.hs (hunk 2 above)
Comment (mightybyte) [reply miner/Miner.hs]: Please no type applications where a simple print statement would work just fine.

Comment (fosskers) [reply miner/Miner.hs]: Not quite the same thing. `logFuncL` is a lens, it's how the ReaderT stack gets its logging "for free".

File: miner/Miner.hs (hunk 5 above)
Comment (fosskers) [reply miner/Miner.hs]: Very easy to add back as a follow-up.

File: chainweb.cabal (hunk 7 above)
Comment (fosskers) [reply chainweb.cabal]: I definitely thought the decision over. Here were my thoughts:

- Chainweb's logging is somewhat complicated, is tailored for Chainweb, is an amalgamation of 4 different module groups (`yet-another-logger` + `loglevel` + `Chainweb.Logger` + `Utils.Logging`), and Lars wants to change it anyway
- `yet-another-logger` states of itself "This Version is yet a preview", Lars tells me it's bit-rotten, it's also a `ReaderT` pattern, and I foresaw it interacting dubiously with the parallelism in the miner
- `rio` is maintained, it's tailored for CLI-based Haskell apps, it has sane defaults, very low boilerplate, easy to use, plays nice with the rest of the ecosystem, I already knew how to use it, and I had a single sitting to come up with something professional

So I went with it. It's certainly a leap above print statements, especially w.r.t. quality of output and configurable log levels. It was effectively a drop-in replacement, save the refactoring I did of `main` itself.

File: miner/Miner.hs (hunk 10 above)
Comment (fosskers) [reply miner/Miner.hs]: I'll comment out (not delete) all the things that Greg will soon need, so that he can jump right in.

**Pull request merged**